# Bams and Sams
###### Tim Tickle, Karthik Shekhar, and Asma Bankapur

*June 2, 2016*

![Tweedledum and Tweedledee](http://wordyenglish.com/alice/i/lg2/lg_18.png)

> "Contrariwise, if it was so, it might be; and if it were so, it would be; but as it isn't, it ain't. That's logic."

> *Tweedledum and Tweedledee*

## Review of relevant commands

`pwd` print working directory

`cd` change directory

`rm` remove file

`rm -r` remove directory

`ls -lh` list contents of directory in **l**ong form, **h**uman-readable format

  > human-readable format displays file sizes in MB, GB, etc. rather than in bytes
  
`ls -a` list all contests in directory, including secret files

  > secret files start with `.`
  
  > `.timsfile.txt` can't see
  
  > `timsfile.txt` can see
  
`cp` copy file

`cp -r` copy directory

`ssh` log into a server

  > **ssh** = **s**ecure **sh**ell
  
  > Can't do anything computationally heavy on the Broad server (ex. Gold)
  
  > **ssh** lets you into the Broad's house while **qrsh** lets you sit at a particular computer and work
  
`qrsh -q interactive` log into the qsub server (a Broad cluster) once on UGER

`>` write to a file

`>>` append to a file

`|` pipe (chain UNIX commands together)

`*` wildcard

`less` opens up a portion of a file in read-only format (**can** see zipped files)

`cat` prints everything to screen, particularly useful when file is input into a pipe (**cannot** see zipped files)

> `zcat` is an updated `cat` command that **can** print zipped files to screen

`mkdir` make directory

`use` use dockit tools present on the Broad server (ex. `use uger`)

`hptmp` scratch space

  > Each Broad user has roughly 2 TB of space (actually goes up to 3-3.5 TB before being locked out)
  
  > Great to run pipelines here
  
  > Files 2 weeks old will be deleted
  
  > Scratch space means that your work here is **not** backed up (ex. think of this space as a digital scratch pad)
  
  > You should move important files into established directories accordingly

## Today's Lesson

##### Login

`ssh <username>@gold`

`use uger`

`qrsh -q interactive`

`mkdir /broad/hptmp/<username>`

`cd /broad/hptmp/<username>`

##### What are BAM and SAM files?

> "Never use a SAM file. Always convert to BAM and walk away."

> *Tim Tickle*

SAM files have a [standardized format](https://samtools.github.io/hts-specs/SAMv1.pdf)

BAM file | SAM file
-------- | --------
Compressed | Not compressed

* BAM/SAM files can be thought of as a more annotated FASTQ file
* Most often, FASTQ reads are aligned against a reference genome and are compressed into a BAM file
  * Many alignment programs exist (ex. BLAST, bowtie, star, kallisto, and more)
  * BAM/SAM files are not always aligned and sometimes might be composed of 2 paired-end reads collapsed into 1 file
* We typically like to only save BAM files, as storage space can be expensive (~ $10,000 / month)

BAM/SAM files have 2 components:

1. Header
2. Alignments

Let's compare the storage space between a SAM and BAM file

`ls -lh /broad/hptmp/ttickle/code_club/test*` We see this SAM file to be ~ 4X more space than the BAM file

> ##### Aside on mutations

> Mutations in reads can (1) biological or due to (2) technical/sequencing errors

> * Alignment programs have gap penalties that help align mismatched sequences
>  * RNA alignment programs typically allow for more breaks/gaps (alternative splicing) than DNA alignment programs
>  * Temple Smith and Michael Waterman came up with the original local sequence alignment algorithm by creating a variation of the Needleman-Wunsch global sequence alignment

A SAM file is text file. Let's `less` it to look at it:

`less -S /broad/hptmp/ttickle/code_club/test.sam` The `-S` parameter stops text-wrapping and allows us to visually see data in columns

`q` to quit the `vim` editor

If we try to `less` the BAM file, we are prompted with the message:

`"/broad/hptmp/ttickle/code_club/test.bam" may be a binary file. See it anyway?` Type `n` for now and we will explore BAM tools later.

### Features and symbols of a SAM file

##### Header

Symbol | Meaning
------ | -------
@ | Header info, **not** an alignment
@HD | Head of first line
VN:1.5 | Version 1.5
SO:coordinate | **s**ort **o**rder by coordinate (chromosome)
@SQ | Reference sequence
SN | Reference sequence name
SN:1 | Chromosome 1 sequence
SN:MT | Mitochondria sequence
SN:GL000* | Additional contigs
LN | Reference length
@RG | Read group
@PG | Alignment program used

##### Alignments

Lines that do not start with `@` represent our actual alignments

Column | Contents
------ | -------
1 | Name of query (read)
2 | Flag description [flags](http://djf604.github.io/SAM-flags-explained-improved/)
3 | Reference name (chromosome #)
4 | First base position on reference
5 | Mapping quality
6 | Cigar string (see below)
10 | Sequence
11 | Base quality scores (FASTQ format)

##### Cigar string legend

Symbol | Meaning
------ | -------
M | Alignment match (sequence match or mismatch)
I | Insertion into reference
D | Deletion from reference
= | Match
X | Mismatch
N | Intronic

How do we see what tools are already on the Broad server?

`use -l`

Let's load a tool that can be used when working with SAM/BAM files.

`use Samtools` can also download from [Samtools](http://samtools.sourceforge.net/)

`samtools --help` and `man samtools` will open up documentation 

The `samtools -h` command does not exist

`samtools view /broad/hptmp/ttickle/code_club/test.bam` will need to `control + c` to stop printing to screen

Instead, let's run the following:

`samtools view /broad/hptmp/ttickle/code_club/test.bam | less -S` notice that `less`brings the output to a stop

If we also want to include header information, include the `-h` parameter:

`samtools view -h /broad/hptmp/ttickle/code_club/test.bam | less -S`

##### Convert a SAM to a BAM

*Include the reference genome, view the SAM file in binary format, and create a new file*

`samtools view -bT /broad/hptmp/ttickle/code_club/hg19.fasta /broad/hptmp/ttickle/code_club/test.sam > converted.bam`

##### Indexing a BAM file

Indexing a file helps speed up computational algorithms (reads are indexed by chromosome #)

`samtools index converted.bam` Creates the indexed BAM file `converted.bam.bai`

`less -S converted.bam.bai` We can't visually see an indexed file because it is in binary format

## Upcoming

1. Sorting
2. Quality control (ex. # of reads mapped)
3. Inspecting coverage and alignments through the command line and IGV
4. How to filter bad reads out
